theatre/studio/src/panels/SequenceEditorPanel/DopeSheet/Right/collectAggregateKeyframes.tsx

```diff
@@ -0,0 +1,118 @@
+import getStudio from '@theatre/studio/getStudio'
+import {val} from '@theatre/dataverse'
+import type {
+  SequenceEditorTree_PropWithChildren,
+  SequenceEditorTree_SheetObject,
+} from '@theatre/studio/panels/SequenceEditorPanel/layout/tree'
+import type {SequenceTrackId} from '@theatre/shared/utils/ids'
+import type {
+  Keyframe,
+  TrackData,
+} from '@theatre/core/projects/store/types/SheetState_Historic'
+import type {IUtilLogger} from '@theatre/shared/logger'
+
+/**
+ * An index over a series of keyframes that have been collected from different tracks.
+ *
+ * Usually constructed via {@link collectAggregateKeyframesInPrism}.
+ */
+export type AggregatedKeyframes = {
+  byPosition: Map<number, KeyframeWithTrack[]>
+  tracks: TrackWithId[]
+}
+
+export type TrackWithId = {
+  id: SequenceTrackId
+  data: TrackData
+}
+
+export type KeyframeWithTrack = {
+  kf: Keyframe
+  track: TrackWithId
+}
+
+/**
+ * Collect {@link AggregatedKeyframes} information from the given tree row with children.
+ *
+ * Must be called within a `prism` context.
+ *
+ * Implementation progress 2/10:
+ *  - This currently does a lot of duplicate work for each compound rows' compound rows.
+ *
+ * Note that we do not need to filter to only tracks that should be displayed, because we
+ * do not do anything counting or iterating over all tracks.
+ */
+export function collectAggregateKeyframesInPrism(
```

Comment:
Why `InPrism`? Couldn't this function just use a prism itself and return a derivation? If not, you can use `prism.ensurePrism()` to ensure this is called in a prism.